File: DL-PR36CTD/DL-PR36CTD (kp=8192,kec=1024).lua

-- https://www.decentlab.com/products/high-precision-pressure-/-liquid-level-temperature-and-electrical-conductivity-sensor-for-lorawan

local PROTOCOL_VERSION = 2

-- device-specific parameters
local PARAMETERS = {
  ["kp"] = 8192,
  ["kec"] = 1024
}

-- sensor definitions
local SENSORS = {
  {["length"] = 4,
   ["values"] = {
     {["name"] = "pressure",
      ["display_name"] = "Pressure",
      ["convert"] = function (x) return (x[0 + 1] - 32768) / PARAMETERS["kp"] end,
      ["unit"] = "bar"},
     {["name"] = "temperature_electronics",
      ["display_name"] = "Temperature (electronics)",
      ["convert"] = function (x) return (x[1 + 1] - 32768) / 256 end,
      ["unit"] = "°C"},
     {["name"] = "temperature_pt1000",
      ["display_name"] = "Temperature (PT1000)",
      ["convert"] = function (x) return (x[2 + 1] - 32768) / 256 end,
      ["unit"] = "°C"},
     {["name"] = "electrical_conductivity",
      ["display_name"] = "Electrical conductivity",
      ["convert"] = function (x) return (x[3 + 1] - 32768) / PARAMETERS["kec"] end,
      ["unit"] = "mS⋅cm⁻¹"}
   }},
  {["length"] = 1,
   ["values"] = {
     {["name"] = "battery_voltage",
      ["display_name"] = "Battery voltage",
      ["convert"] = function (x) return x[0 + 1] / 1000 end,
      ["unit"] = "V"}
   }}
}

-- helper functions
local function fromhex(s)
  local arr = {}
  local k = 1
  for i = 1, #s, 2 do
    arr[k] = tonumber(s:sub(i, i + 1), 16)
    k = k + 1
  end
  return arr
end

local function toint(b1, b2)
  return b1 * 256 + b2
end


-- decoding function
local function decentlab_decode(msg)
  local bytes = msg
  if type(msg) == "string" then
    bytes = fromhex(msg)
  end

  local version = bytes[1]
  if version ~= PROTOCOL_VERSION then
    error("protocol version " .. version .. " doesn't match v2")
  end

  local device_id = toint(bytes[2], bytes[3])
  local flags = toint(bytes[4], bytes[5])
  local result = {["device_id"] = device_id, ["protocol_version"] = version}
  local k = 6
  -- decode sensors
  for _, sensor in ipairs(SENSORS) do
    if flags % 2 == 1 then
      local x = {}
      for j = 1, sensor["length"] do
        x[#x + 1] = toint(bytes[k], bytes[k + 1])
        k = k + 2
      end

      -- decode sensor values
      for _, value in ipairs(sensor["values"]) do
        if value["convert"] then
          result[value["name"]] = {
            ["value"] = value["convert"](x),
            ["display_name"] = value["display_name"],
            ["unit"] = value["unit"]
          }
        end -- if sensor value used
      end -- for each sensor value
    end -- if sensor values present
    flags = math.floor(flags / 2)
  end -- for each sensor

  return result
end


-- test
local payloads = {
  "020a17000380079786978180060c2b",
  "020a1700020c2b",
}

local function main()
  for _, pl in ipairs(payloads) do
    local decoded = decentlab_decode(pl)
    for k, v in pairs(decoded) do
      print(k .. ": " .. (type(v) == "table" and (v["value"] .. " " .. (v["unit"]  or "")) or v))
    end
    print()
  end
end

main()

File: DL-PR36CTD/DL-PR36CTD (kp=8192,kec=1024).php
<?php
/* https://www.decentlab.com/products/high-precision-pressure-/-liquid-level-temperature-and-electrical-conductivity-sensor-for-lorawan */

abstract class DecentlabDecoder
{
    const PROTOCOL_VERSION = 2;

    protected $sensors;

    public function decode($payload = '')
    {
        $this->bytes = hex2bin($payload);
        $this->parts = [];

        $this->parts['version'] = ord($this->bytes[0]);
        if ($this->parts['version'] != self::PROTOCOL_VERSION) {
            $this->parts['error'] = sprintf("protocol version %u doesn't match v2", $this->parts['version']);
            return $this->parts;
        }

        $this->parts['device_id'] = unpack('n', substr($this->bytes, 1))[1];
        $flags = unpack('n', substr($this->bytes, 3))[1];

        /* decode payload */
        $k = 5;
        foreach ($this->sensors as $sensor) {
            if (($flags & 1) == 1) {
                $x = [];
                /* convert data to 16-bit integer array */
                for ($j = 0; $j < $sensor['length']; $j++) {
                    array_push($x, unpack('n', substr($this->bytes, $k))[1]);
                    $k += 2;
                }

                /* decode sensor values */
                foreach ($sensor['values'] as $value) {
                    if ($value['convert'] != NULL) {
                        $this->parts[$value['name'] . '_value'] = $value['convert']($x);
                        if ($value['unit'] != NULL) {
                            $this->parts[$value['name'] . '_unit'] = $value['unit'];
                        }
                    }
                }
            }
            $flags >>= 1;
        }

        return $this->parts;
    }
}

class DL_PR36CTD_Decoder extends DecentlabDecoder {
    /* device-specific parameters */
    public function __construct($kp, $kec)
    {
        $this->sensors = [
            [
                'length' => 4,
                'values' => [
                    [
                        'name' => 'pressure',
                        'convert' => function ($x) use ($kp, $kec) { return ($x[0] - 32768) / $kp; },
                        'unit' => 'bar',
                    ],
                    [
                        'name' => 'temperature_electronics',
                        'convert' => function ($x) use ($kp, $kec) { return ($x[1] - 32768) / 256; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'temperature_pt1000',
                        'convert' => function ($x) use ($kp, $kec) { return ($x[2] - 32768) / 256; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'electrical_conductivity',
                        'convert' => function ($x) use ($kp, $kec) { return ($x[3] - 32768) / $kec; },
                        'unit' => 'mS⋅cm⁻¹',
                    ],
                ],
            ],
            [
                'length' => 1,
                'values' => [
                    [
                        'name' => 'battery_voltage',
                        'convert' => function ($x) use ($kp, $kec) { return $x[0] / 1000; },
                        'unit' => 'V',
                    ],
                ],
            ],
        ];
    }
}


$decoder = new DL_PR36CTD_Decoder(8192, 1024);
$payloads = [
    '020a17000380079786978180060c2b',
    '020a1700020c2b',
];

foreach($payloads as $payload) {
    var_dump($decoder->decode($payload));
}
